src/seabreeze/pyseabreeze/transport.py

```diff
@@ -364,3 +368,254 @@ def get_name_from_pyusb_backend(backend: usb.backend.IBackend) -> str | None:
+            size = 64
+        if kwargs:
+            warnings.warn(f"kwargs provided but ignored: {kwargs}")
+        if timeout_ms:
+            self._device.socket.settimeout(timeout_ms / 1000.0)
+        data = bytearray(size)
+        toread = size
+        view = memoryview(data)
+        while toread:
+            nbytes = self._device.socket.recv_into(view, toread)
+            view = view[nbytes:]
+            toread -= nbytes
+        return bytes(data)
+
+    @property
+    def default_timeout_ms(self) -> int:
+        if not self._device:
+            raise RuntimeError("no protocol instance available")
+        timeout = self._device.socket.gettimeout()
+        if not timeout:
+            return 10000
+        return int(timeout * 1000)
+
+    @property
+    def protocol(self) -> PySeaBreezeProtocol:
+        if self._protocol is None:
+            raise RuntimeError("no protocol instance available")
+        return self._protocol
+
+    @classmethod
+    def list_devices(cls, **kwargs: Any) -> Iterable[IPv4TransportHandle]:
+        """list IPv4 devices for all available spectrometers
+
+        Note: this includes spectrometers that are currently opened in other
+        processes on the machine.
+
+        Yields
+        ------
+        devices : IPv4TransportHandle
+            unique socket devices for each available spectrometer
+        """
+        dev_sockets = []
+
+        # Use multicast to discover potential spectrometers. Requires a network
+        # adapter to be specified.
+        network_adapter = kwargs.get("network_adapter", None)
+        if network_adapter:
+            multicast_group = (
+                kwargs.get("multicast_group", "239.239.239.239"),
+                kwargs.get("multicast_port", 57357),
+            )  # default values for HDX devices
+            # Create the datagram (UDP) socket
+            sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+            # Set a timeout so the socket does not block
+            # indefinitely when trying to receive data.
+            sock.settimeout(kwargs.get("multicast_timeout", 2))
+            sock.setsockopt(
+                socket.IPPROTO_IP,
+                socket.IP_MULTICAST_IF,
+                socket.inet_aton(network_adapter),
+            )
+            # prepare a message requesting all devices in the multicast group
+            # to send their (USB) product id
+            transport = IPv4Transport(OBPProtocol)
+            protocol = OBPProtocol(transport)
+            msg_type = 0xE01  # Product ID
+            data = protocol.msgs[msg_type](*())
+            message = protocol._construct_outgoing_message(
+                msg_type,
+                data,
+                request_ack=True,
+            )
+            sock.sendto(message, multicast_group)
+            print("Waiting to receive multicast response(s)")
+            while True:
+                try:
+                    data = bytearray(90)
```

Comment:
No, this was mostly to leave space above the 64 bytes for a potential payload. I did experiment with different commands, originally requesting the serial number. I was not quite sure whether that would be sent via payload or not. So 90 bytes is just to leave room and otherwise arbitrary.

I think that `0xE01` is always sent in the "immediate data" instead for as payload..? The specs I have are, unfortunately, not very clear in this regard.